ipn/ipnlocal/local.go

```diff
@@ -266,6 +266,15 @@ type LocalBackend struct {
 	// at the moment that tkaSyncLock is taken).
 	tkaSyncLock sync.Mutex
 	clock       tstime.Clock
+
+	// c2nUpdateStatus is the status of c2n-triggered client update, protected
+	// by c2nUpdateStatusMu.
+	c2nUpdateStatusMu sync.Mutex
```

Comment:
Ideally we wouldn't have more than one Mutex in this type. I think statusLock and tkaSyncLock shouldn't even exist.

Can you use the `mu` lock instead?

Otherwise you need to document the locking order between b.mu and this one.

Reply:
Oh, you run `exec.Command(cmdTS, "version", "--json").Output()` while holding the mutex. That's not ideal to do while holding LocalBackend.mu. So you'd need to release the mutex earlier, without `defer b.c2nUpdateStatusMu.Unlock()` , and then be sure to clear it back to false in any error path, like:

```go
   exec.Command(cmdTS, "version", "--json").Output()
  ....
   defer func() {
         if ret.Err != nil { 
               b.setC2NUpdateRunning(false) // (and this would do the lines 207-208)
        }
   }()
```

Reply:
Ok, switched to `mu` and reorganized things a bit to avoid races.
Out of curiosity: we don't like multiple mutexes for independent data because the locking order can be confusing?

Reply:
Because we've had too many deadlocks in the past with multiple mutexes. It not only requires a lot more documentation but it also requires more brainpower to follow the rules which empirically we're not smart enough about doing. And/or we don't have good static analysis tooling around it. So in lieu of tooling or brain, we prefer to keep it simple so it's less likely to be screwed up. It also means we can name methods "doFooLocked" with that "Locked" suffix and it's obvious which lock you mean: the only one.